enforcer/src/keystate/keystate_list_cmd.c

```diff
@@ -334,48 +236,245 @@ handles(const char *cmd, ssize_t n)
+            zone_name(zone),
+            key_data_role_text(key),
+            map_keystate(key),
+            tchange,
+            hsm_key_bits(hsmkey),
+            hsm_key_algorithm(hsmkey),
+            hsm_key_locator(hsmkey),
+            hsm_key_repository(hsmkey),
+            key_data_keytag(key));
+}
+
+static void
+printdebugheader(int sockfd) {
+    client_printf(sockfd,
+            "Keys:\nZone:                           Key role:     "
+            "DS:          DNSKEY:      RRSIGDNSKEY: RRSIG:       "
+            "Pub: Act: Id:\n");
+}
+
+static void
+printdebugkey(int sockfd, zone_t* zone, key_data_t* key, char* tchange, hsm_key_t* hsmkey) {
+    (void)tchange;
+    client_printf(sockfd,
+            "%-31s %-13s %-12s %-12s %-12s %-12s %d %4d    %s\n",
+            zone_name(zone),
+            key_data_role_text(key),
+            key_state_state_text(key_data_cached_ds(key)),
+            key_state_state_text(key_data_cached_dnskey(key)),
+            key_state_state_text(key_data_cached_rrsigdnskey(key)),
+            key_state_state_text(key_data_cached_rrsig(key)),
+            key_data_publish(key),
+            key_data_active_ksk(key) | key_data_active_zsk(key),
+            hsm_key_locator(hsmkey));
+}
+
+static void
+printdebugparsablekey(int sockfd, zone_t* zone, key_data_t* key, char* tchange, hsm_key_t* hsmkey) {
+    (void)tchange;
+    client_printf(sockfd,
+            "%s;%s;%s;%s;%s;%s;%d;%d;%s\n",
+            zone_name(zone),
+            key_data_role_text(key),
+            key_state_state_text(key_data_cached_ds(key)),
+            key_state_state_text(key_data_cached_dnskey(key)),
+            key_state_state_text(key_data_cached_rrsigdnskey(key)),
+            key_state_state_text(key_data_cached_rrsig(key)),
+            key_data_publish(key),
+            key_data_active_ksk(key) | key_data_active_zsk(key),
+            hsm_key_locator(hsmkey));
+}
+
+static char **
+tokenizeparam(char *argument)
+{
+    char** tokenized;
+    char** newtokenized;
+    int argCount, i;
+    char* argString;
+    char* argSavePtr;
+    int argSize = 8;
+
+    if ((argString = strtok_r(argument, ",", &argSavePtr)) != NULL) {
+        if ((tokenized = malloc(sizeof (char*)*argSize)) == NULL) {
+            return NULL;
+        }
+        argCount = 0;
+        do {
+            if (strcmp(argString, "")) {
+                tokenized[argCount] = argString;
+                ++argCount;
+                if (argCount == argSize) {
+                    argSize *= 2;
+                    if ((newtokenized = realloc(tokenized, sizeof (char*)*argSize)) == NULL) {
+                        free(tokenized);
+                        return NULL;
+                    }
+                    tokenized = newtokenized;
+                }
+            }
+        } while (strtok_r(NULL, ",", &argSavePtr) != NULL);
+        tokenized[argCount] = NULL;
+    } else {
+        if ((tokenized = malloc(sizeof (char*)*argSize)) == NULL) {
```

Comment:
I find the guarded-free a good pattern to use.  In case you need to explicitly close or dispose the item, the if is required and it it helpful to remind you the object may not yet be initialized.  This patterb is actually needed in OO languages and good for C as well for the same reason.
The argSize is a bit of nit picking, the handful of bytes that will additionally be alloced over the period of a millisecond will not eat up your memory.
The additional free on the bAll is also good.  Note that the test if the -a and -e are not both given on the command line was not there earlier, and this would actually be required.  Most programs would allow you to override a parameter such as -e and -a, or give one precedence.  It is only for historical reasons not to allow is now, and keeping the test will help keeping the code stable.
It is like the way-to-many argument checking all over the code, there are not directly related code, this is contract keeping.
As a side note, it is far more logical for the -e parameter to accept an argument "all", rather than to have a separate -a options.